src/components/modals/FlipInputModal2.tsx

```diff
@@ -0,0 +1,301 @@
+import { formatNumber } from '../../locales/intl'
+import s from '../../locales/strings'
+import { deviceHeight } from '../../theme/variables/platform'
+import { useSelector } from '../../types/reactRedux'
+import { getCurrencyCode } from '../../util/CurrencyInfoHelpers'
+import { DECIMAL_PRECISION } from '../../util/utils'
+import { Card } from '../cards/Card'
+import { ExchangeRate2 } from '../common/ExchangeRate2'
+import { cacheStyles, Theme, useTheme } from '../services/ThemeContext'
+import { FiatText } from '../text/FiatText'
+import { EdgeText } from '../themed/EdgeText'
+import { ExchangedFlipInput2, ExchangedFlipInputAmounts, ExchangedFlipInputRef, ExchangeFlipInputFields } from '../themed/ExchangedFlipInput2'
+import { MiniButton } from '../themed/MiniButton'
+import { ThemedModal } from '../themed/ThemedModal'
+
+export type FlipInputModalResult = {
+  nativeAmount: string
+  exchangeAmount: string
+}
+
+export interface SetFeesParams {
+  feeTokenId?: string
+  feeNativeAmount: string
+}
+export interface FlipInputModalRef {
+  setFees: (params: SetFeesParams) => void
+  setAmount: (field: ExchangeFlipInputFields, value: string) => void
+  setError: (errorMessage: string | null) => void
+}
+
+type FeeStyleTypes = 'dangerText' | 'warningText'
+
+type Props = {
+  bridge: AirshipBridge<FlipInputModalResult>
+  walletId: string
+  tokenId?: string
+  startNativeAmount?: string
+  forceField?: ExchangeFlipInputFields
+  // Fees
+  feeTokenId?: string
+  feeNativeAmount?: string
+  feeStyle?: FeeStyleTypes
+  onFeesChange?: () => void
+  onMaxSet?: () => void
+  onAmountsChanged?: (amounts: ExchangedFlipInputAmounts) => void
+  headerText?: string
+  hideMaxButton?: boolean
+}
+
+const FlipInputModal2Component = React.forwardRef<FlipInputModalRef, Props>((props: Props, ref) => {
+  const {
+    bridge,
+    walletId,
+    tokenId,
+    startNativeAmount,
+    forceField,
+    onAmountsChanged,
+    feeTokenId: startingFeeTokenId,
+    feeNativeAmount: startingFeeNativeAmount = '',
+    feeStyle,
+    onFeesChange,
+    onMaxSet,
+    headerText,
+    hideMaxButton
+  } = props
+
+  const account = useSelector(state => state.core.account)
+  const currencyWallets = useWatch(account, 'currencyWallets')
+  const [feeTokenId, setFeeTokenId] = useState<string | undefined>(startingFeeTokenId)
+  const [feeNativeAmount, setFeeNativeAmount] = useState<string>(startingFeeNativeAmount)
+  const [errorMessage, setErrorMessage] = useState<string | null>(null)
+  const wallet = currencyWallets[walletId]
+  const { currencyInfo } = wallet
+  const balances = useWatch(wallet, 'balances')
+  const { pluginId } = currencyInfo
+  const currencyCode = getCurrencyCode(wallet, tokenId)
+  const displayDenom = useDisplayDenom(pluginId, currencyCode)
+  const exchangedFlipInputRef = React.useRef<ExchangedFlipInputRef>(null)
+  const feeCurrencyCode = getCurrencyCode(wallet, feeTokenId)
+  const feeDisplayDenom = useDisplayDenom(pluginId, feeCurrencyCode)
```

Comment:
Optional cleanup: Can we separate the `useState` hooks from the others, which are more subscriptions? That way it's more clear what we depend on vs. what we store.

Reply:
Ok. did my best but many of the lines are dependent on each other forcing a particular order